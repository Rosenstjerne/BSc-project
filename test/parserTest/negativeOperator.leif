var int a,b,c;
a = 5;
b = (-a);
c = -b;
print(b); # should print -5
print(c); # should print 5

