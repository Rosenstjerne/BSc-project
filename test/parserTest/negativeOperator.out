
.data
mmapEr:
                .string "mmap failed\n" # error message string for mmap

form:
                .string "%d\n"          # form string for C printf

formTrue:
                .string "true\n"        # form string for C printf

formFalse:
                .string "false\n"       # form string for C printf

.text

.globl main

main:

                                        # CALLEE PROLOGUE
                pushq %rbp              # save caller's base pointer
                movq %rsp, %rbp         # make stack pointer new base pointer

                addq $-24, %rsp         # allocate space for local variables


                movq %rsp, %rbx         # Saves stack pointer in %rbx
                addq $-16, %rsp         # allocate space for extra intermediate
                                        # stack memory

                pushq %r8               # %r8 is callee save
                pushq %r9               # %r9 is callee save
                pushq %r10              # %r10 is callee save
                pushq %r11              # %r11 is callee save
                pushq %r12              # %r12 is callee save
                pushq %r13              # %r13 is callee save
                pushq %r14              # %r14 is callee save
                pushq %r15              # %r15 is callee save

                movq $0, -24(%rbp)      # Initializes c to 0
                movq $0, -16(%rbp)      # Initializes b to 0
                movq $0, -8(%rbp)       # Initializes a to 0
                movq $5, %r9            # Moves integer into reg1
                movq %rbp, %rdx         # preparing for static link computation
                movq %r9, -8(%rdx)      # Move param reg1 into ['fun_1_main',
                                        # 'var_0_a'] (a)
                movq %rbp, %rdx         # preparing for static link computation
                movq -8(%rdx), %r11     # Move param var_0_a (a) into reg3
                movq %r11, %rcx         # Moves reg3 out into %rax for
                                        # operation
                movq $0, %rax           # Move reg3 to reg4
                subq %rcx, %rax         # Operation: $0 - reg4
                movq %rax, %r12         # Moves result into reg4
                movq %rbp, %rdx         # preparing for static link computation
                movq %r12, -16(%rdx)    # Move param reg4 into ['fun_1_main',
                                        # 'var_1_b'] (b)
                movq %rbp, %rdx         # preparing for static link computation
                movq -16(%rdx), %r14    # Move param var_1_b (b) into reg6
                movq %r14, %rcx         # Moves reg6 out into %rax for
                                        # operation
                movq $0, %rax           # Move reg6 to reg7
                subq %rcx, %rax         # Operation: $0 - reg7
                movq %rax, %r15         # Moves result into reg7
                movq %rbp, %rdx         # preparing for static link computation
                movq %r15, -24(%rdx)    # Move param reg7 into ['fun_1_main',
                                        # 'var_2_c'] (c)
                movq %rbp, %rdx         # preparing for static link computation
                movq -16(%rdx), %rcx    # 
                movq %rcx, 0(%rbx)      # Move param var_1_b (b) into reg8
                movq 0(%rbx), %rcx      # Moves reg8 out of the stack for print
                                        # PRINTING
                leaq form(%rip), %rdi   # pass 1. argument in %rdi
                movq %rcx, %rsi         # Moves printable object to rsi
                xorq %rax, %rax         # No floating point arguments

                callq printf@plt        # calls the printf method
                                        # END OF PRINTING
                movq %rbp, %rdx         # preparing for static link computation
                movq -24(%rdx), %rcx    # 
                movq %rcx, -8(%rbx)     # Move param var_2_c (c) into reg9
                movq -8(%rbx), %rcx     # Moves reg9 out of the stack for print
                                        # PRINTING
                leaq form(%rip), %rdi   # pass 1. argument in %rdi
                movq %rcx, %rsi         # Moves printable object to rsi
                xorq %rax, %rax         # No floating point arguments

                callq printf@plt        # calls the printf method
                                        # END OF PRINTING
end_main:

                popq %r15               # restore callee save register %r15
                popq %r14               # restore callee save register %r14
                popq %r13               # restore callee save register %r13
                popq %r12               # restore callee save register %r12
                popq %r11               # restore callee save register %r11
                popq %r10               # restore callee save register %r10
                popq %r9                # restore callee save register %r9
                popq %r8                # restore callee save register %r8

                addq $16, %rsp          # deallocate stack space for extra
                                        # intermediate stack memory


                                        # CALLEE EPILOGUE
                movq %rbp, %rsp         # restore stack pointer
                popq %rbp               # restore base pointer
                ret                     # return from call

mmap_error:
                leaq mmapEr(%rip), %rdi # pass 1. arg in %rdi
                xor %rax, %rax          # no floating point arguments
                callq printf            # calls printf method

                movq $60, %rax          # prepare for exit syscall
                xor %rdi, %rdi          # exit status 0
                syscall                 # exit


